doc/tutorials/12-constant_pH/12-constant_pH.ipynb

```diff
@@ -166,95 +297,53 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "num_samples = 100\n",
+    "# the reference data from Henderson-Hasselbalch equation\n",
+    "def ideal_alpha(pH, pK):\n",
+    "    return 1. / (1 + 10**(pK - pH))\n",
     "\n",
-    "pHs = np.linspace(1, 8, num=15)\n",
-    "degrees_of_dissociation = []\n",
-    "std_dev_degree_of_dissociation = []\n",
-    "histograms = []\n",
-    "histogram_edges = range(0,N0+1)\n",
+    "# empty lists as placeholders for collecting data\n",
+    "numAs = []; # number of A- species observed at each sample\n",
     "\n",
+    "#run a productive simulation and collect the data\n",
+    "print(\"Simulated pH values: \",pHs)\n",
     "for pH in pHs:\n",
-    "    print(\"pH is {:.1f}\".format(pH))\n",
+    "    print(\"Run pH {:.2f} ...\".format(pH))\n",
     "    RE.constant_pH = pH\n",
-    "    RE.reaction(4 * N0) # equilibration to the new pH\n",
-    "    numAs = []\n",
+    "    numAs_current = []; # temporary data storage for a given pH\n",
+    "    RE.reaction(20*N_acid + 1) # pre-equilibrate to the new pH value\n",
     "    for i in range(num_samples):\n",
-    "        RE.reaction(N0 + 1)\n",
-    "        numAs.append(system.number_of_particles(type=type_A))\n",
-    "    degrees_of_dissociation.append(np.mean(numAs) / N0)\n",
-    "    std_dev_degree_of_dissociation.append(np.std(numAs, ddof=1) / N0)\n",
-    "    print(\"occurred particle numbers of type A \", sorted(set(numAs)))\n",
-    "    histogram = np.bincount(np.array(numAs).astype(int), minlength=N0 + 1)\n",
-    "    histograms.append(histogram / float(len(numAs)))\n",
-    "    if(abs(pH - 1.0) < 1e-4):\n",
-    "        print(\"acceptance rate dissociation\", RE.get_acceptance_rate_reaction(0))\n",
-    "        print(\"acceptance rate association\", RE.get_acceptance_rate_reaction(1))"
+    "        if(np.random.random()<PROB_REACTION): \n",
```

Comment:
I don't think this is needed. You can run MD and MC moves sequentially. This is normal practice. Sequentiallity will only affect how fast one converges to equilibrium. In the limit of infinite runtime the exact sequence of MC/MD moves does not matter anymore. https://link.springer.com/chapter/10.1007/978-3-642-41315-5_23 See figure 1.

Reply:
It is common practice but I am not aware of any proof that it is correct. In general, performing MC moves sequentially instead of a random order breaks detailed balance. If there is enough randomness in the system, this breaking might be weak and hardly noticeable. Sot it is not advisable to do it like that.

Even if we routinely break this rule, we should not teach others to do so.

Reply:
I am not aware of a proof on how detailed balance is broken by sequential MC/MD moves.

Reply:
Some years back, Olaf Lenz was showing an example of that at the ICP seminar.  I cannot immediately reproduce it now.

Are you aware that a random order could break detailed balance? If not, then we should keep the random order as a safe choice.

Reply:
I challenge that sequentially using MD and then MC moves breaks sampling from the correct probabiliy distribution for the ensemble at hand :
The paper "A rigorous procedure for combining molecular dynamics and Monte Carlo simulation algorithms" demonstrates that the correct probability distribution is sampled for the canonical (NVT) case:
https://www.sciencedirect.com/science/article/pii/S0301010400002469

They perform MDMC steps. One MDMC step consists of 

1. m MC steps
2. integrating n timesteps with MD

In section 3 they show that this is a valid approach.

Reply:
The problem is that you are proposing a sequential MC scheme. While this can be correct, it can also fail. See the discussion in [1]:

> The implications of our theorem are significant. First, the sequential updating approach, widely practiced and nearly as widely criticized, is correct. All that is required is that the
sweep leave the Boltzmann distribution invariant, in addition to leading to regular sampling. Indeed, there are numerical reports that the sequential updating approach leads not only
to correct sampling but also to more efficient sampling in a variety of contexts, ranging from image processing to membrane physics. There are some simple cases, however, for
which random Metropolis updating is ergodic and regular, but sequential Metropolis updating is not. One is the Ising model at infinite temperature, where the configurations produced by sequential updating simply oscillate between the initial configuration and the configuration with all spins flipped. Another example is the 2x2 Ising model at any temperature. If the spins are updated in the order [[1 2][3 4]], then the initial configurations [[+ -][+ +]] and [[+ +][- +]] lead to sequences that also oscillate between spin–flipped configurations. In these examples, our theorem does not guarantee convergence to the Boltzmann distribution, because assumption (4) is violated. If, however, the individual spin–flip moves within the sweep were performed with a probability (1 - a), as in Eq. (5), then convergence to the Boltzmann distribution would occur in both of these examples.

This shows that sequential updating can break the MC simulation. Therefore, we should be careful how we teach it.

The paper you cite they show that you can combine configurational MC with MD. But they do not discuss whether this is true when you combine "orthogonal" algorithms like cpH and MD (where one samples the configurational space and the other the composition space).

It might well be that the sequential updating of the cpH and the MD move leads to correct sampling. But what is the advantage to do this in this tutorial? I do not see any real advantage, but the big disadvantage that we teach an advanced method which is not generally valid. If we want to use the sequential updating, then we should discuss why we want use it, why it works and when does not work. All of this is out of scope of this tutorial.

The random mixing of the moves is the most basic way how to do it, and in this tutorial we should stick to be basics when it comes to such subjects.

[1] Manousiouthakis, V. I.; Deem, M. W. Strict Detailed Balance Is Unnecessary in Monte Carlo Simulation. The Journal of Chemical Physics 1999, 110 (6), 2753–2756. https://doi.org/10.1063/1.477973.